Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailClipmapGenerationPass.azsl

```diff
@@ -36,116 +36,96 @@ void MainCS(
     uint numberOfGroups = TerrainSrg::m_clipmapData.m_detailDispatchGroupCountX * TerrainSrg::m_clipmapData.m_detailDispatchGroupCountY;
     uint numberOfGroupThreads = THREAD_NUM_X * THREAD_NUM_Y;
 
-    for (uint clipmapLevel = 0; clipmapLevel < TerrainSrg::m_clipmapData.m_detailClipmapStackSize; ++clipmapLevel)
+    for (uint regionIndex = 0; regionIndex < TerrainSrg::m_clipmapData.m_detailClipmapUpdateRegionCount; ++regionIndex)
     {
-        for (uint updateRegionIndex = 0; updateRegionIndex < UpdateRegionMax; ++updateRegionIndex)
+        TerrainSrg::ClipmapUpdateRegion updateRegion = TerrainSrg::m_detailClipmapUpdateRegions[regionIndex];
+        uint clipmapLevel = updateRegion.m_clipmapLevel;
+        uint4 aabb = updateRegion.m_clipmapRegion;
+        uint2 size = uint2(aabb.z - aabb.x, aabb.w - aabb.y);
+        uint totalNumberOfTexels = size.x * size.y;
+
+        uint totalStride = (totalNumberOfTexels + numberOfGroups - 1) / numberOfGroups;
+        uint totalBegin = min(totalStride * (groupID.x + groupID.y * TerrainSrg::m_clipmapData.m_detailDispatchGroupCountX), totalNumberOfTexels);
+        uint totalEnd = min(totalBegin + totalStride, totalNumberOfTexels);
+
+        uint groupTotal = totalEnd - totalBegin;
+        uint groupStride = (groupTotal + numberOfGroupThreads - 1) / numberOfGroupThreads;
+        uint groupBegin = min(groupStride * (groupThreadID.x + groupThreadID.y * THREAD_NUM_X), groupTotal);
+        uint groupEnd = min(groupBegin + groupStride, groupTotal);
+
+        for (uint i = groupBegin; i < groupEnd; ++i)
```

Comment:
Just checking - is it possible for there to be extra threads at the end left over with nothing to do if the thread count doesn't perfectly divide the work? I'd think they would need to be ignored somehow - possible I'm missing something though.

Reply:
Yes, it can be. They are not always divided perfectly, the last few groups will have equal or fewer texels to update. So they may be wasted when update region is small. But they will gain benefits when we have large areas. Especially when we now have the margin that accumulate small regions.

Reply:
We can actually update this to distribute all the jobs once instead of twice, which can also improve the leftover conditions. I will make some the change later.